extension/encoding/awslogsencodingextension/internal/unmarshaler/s3-access-log/unmarshaler.go

```diff
@@ -0,0 +1,244 @@
+	record := plog.NewLogRecord()
+
+	remaining := log
+	var i int
+	var value string
+	var err error
+	for i = 0; remaining != ""; i++ {
+		if i >= len(attributeNames) {
+			return errors.New("values in log line exceed the number of available fields")
+		}
+
+		value, remaining, err = scanField(remaining)
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return err
+		}
+
+		if value == unknownField && i != fieldIndexACLRequired {
+			// acl required field can be '-' to indicate that no ACL was required
+			continue
+		}
+
+		if i == fieldIndexTime {
+			// This is the timestamp that follows a strict format
+			// "[DD/MM/YYYY:HH:mm:ss zone]". Since zone is after a
+			// space, we cut the string again to remove the zone.
+			// Zone is always UTC, so it will always be +0000.
+			_, remaining, _ = strings.Cut(remaining, " ")
+		}
+
+		if err = addField(i, value, resourceAttr, record); err != nil {
+			return err
+		}
+	}
+
+	if i != fieldIndexACLRequired+1 {
+		return errors.New("values in log line are less than the number of available fields")
+	}
+
+	rScope := scopeLogs.LogRecords().AppendEmpty()
+	record.MoveTo(rScope)
+
+	return nil
+}
+
+func addField(field int, value string, resourceAttr *resourceAttributes, record plog.LogRecord) error {
+	switch field {
+	case fieldIndexS3BucketOwner:
+		resourceAttr.bucketOwner = value
+	case fieldIndexS3BucketName:
+		resourceAttr.bucketName = value
+	case fieldIndexTime:
+		// The format in S3 access logs at this point is as "[DD/MM/YYYY:HH:mm:ss".
+		if value == "" {
+			return errors.New("unexpected: time value is empty")
+		}
+		t, err := time.Parse("02/Jan/2006:15:04:05", value[1:])
+		if err != nil {
+			return fmt.Errorf("failed to get timestamp of log: %w", err)
+		}
+		record.SetTimestamp(pcommon.NewTimestampFromTime(t))
+	case fieldIndexHTTPStatus,
+		fieldIndexBytesSent,
+		fieldIndexObjectSize,
+		fieldIndexTurnAroundTime,
+		fieldIndexTotalTime:
+		n, err := strconv.ParseInt(value, 10, 64)
+		if err != nil {
+			return fmt.Errorf("value for field %q in log line is not a number", field)
+		}
+		attrName := attributeNames[field]
+		record.Attributes().PutInt(attrName, n)
+	case fieldIndexACLRequired:
+		attrName := attributeNames[field]
+		// If the request required an ACL for authorization, the string is Yes.
+		// If no ACLs were required, the string is -.
+		switch value {
+		case "Yes":
+			record.Attributes().PutBool(attrName, true)
+		case unknownField:
+			record.Attributes().PutBool(attrName, false)
+		default:
+			return fmt.Errorf("unknown value %q for field %q", value, field)
+		}
+	case fieldIndexTLSVersion:
+		// The value is one of following: TLSv1.1, TLSv1.2, TLSv1.3.
+		// We get the version after "v".
+		_, remaining, _ := strings.Cut(value, "v")
+		if remaining == "" {
+			_, remaining, _ = strings.Cut(value, "V")
+			if remaining == "" {
+				return fmt.Errorf("missing TLS version: %q", value)
+			}
+		}
```

Comment:
This doesn't look right: we're doing the exact same operation twice aren't we?

Also, should we perhaps use `strings.CutPrefix(value, "TLSv")`?

Reply:
Ah, sorry, earlier I missed that the second one was uppercase V. Then I suggest we look for the `1` rather than the character that comes before it, as otherwise we'll be scanning through the string twice.

```go
i := strings.IndexRune(value, '1')
if i == -1 {
    return fmt.Errorf("missing TLS version: %q", value)
}
tlsVersion := value[i:]
```

Reply:
I love these suggestions!! I have changed it.